File: skills/suno-song-creator/utils/count-prompt.js
#!/usr/bin/env node

/**
 * Suno Prompt Character Counter
 *
 * Accurately counts characters in a Suno prompt to ensure it stays under 1000 characters.
 * Usage: node count-prompt.js "your prompt text here"
 * Or: echo "your prompt text" | node count-prompt.js
 */

const fs = require('fs');

function countPrompt(text) {
  // Remove any leading/trailing whitespace
  const trimmed = text.trim();

  // Count total characters
  const totalChars = trimmed.length;

  // Count lines (sections)
  const lines = trimmed.split('\n').filter(line => line.trim().length > 0);
  const lineCount = lines.length;

  // Calculate remaining characters
  const limit = 1000;
  const remaining = limit - totalChars;

  return {
    totalChars,
    lineCount,
    remaining,
    isValid: totalChars <= limit,
    percentage: ((totalChars / limit) * 100).toFixed(1)
  };
}

function formatOutput(stats) {
  const statusEmoji = stats.isValid ? '✓' : '✗';
  const status = stats.isValid ? 'VALID' : 'EXCEEDS LIMIT';

  console.log('\n' + '='.repeat(50));
  console.log('  Suno Prompt Character Count');
  console.log('='.repeat(50));
  console.log(`\nTotal Characters: ${stats.totalChars}/1000`);
  console.log(`Sections (lines): ${stats.lineCount}`);
  console.log(`Remaining: ${stats.remaining} characters`);
  console.log(`Usage: ${stats.percentage}%`);
  console.log(`\nStatus: ${statusEmoji} ${status}`);

  if (!stats.isValid) {
    console.log(`\n⚠️  OVER LIMIT BY ${Math.abs(stats.remaining)} CHARACTERS`);
    console.log('\nSuggestions:');
    console.log('  • Remove redundant adjectives');
    console.log('  • Use commas instead of "and"');
    console.log('  • Condense longer phrases');
    console.log('  • Remove least essential mood descriptors');
  }

  console.log('\n' + '='.repeat(50) + '\n');

  return stats.isValid ? 0 : 1;
}

// Main execution
let input = '';

if (process.argv[2]) {
  // Command line argument
  input = process.argv[2];
  const stats = countPrompt(input);
  process.exit(formatOutput(stats));
} else if (!process.stdin.isTTY) {
  // Piped input
  process.stdin.setEncoding('utf8');

  process.stdin.on('data', (chunk) => {
    input += chunk;
  });

  process.stdin.on('end', () => {
    const stats = countPrompt(input);
    process.exit(formatOutput(stats));
  });
} else {
  console.error('Usage: node count-prompt.js "your prompt text"');
  console.error('   Or: echo "your prompt text" | node count-prompt.js');
  process.exit(1);
}

File: skills/suno-song-creator/utils/count-prompt.py
#!/usr/bin/env python3

"""
Suno Prompt Character Counter

Accurately counts characters in a Suno prompt to ensure it stays under 1000 characters.
Usage: python count-prompt.py "your prompt text here"
Or: echo "your prompt text" | python count-prompt.py
"""

import sys


def count_prompt(text):
    """Count characters in a Suno prompt."""
    # Remove leading/trailing whitespace
    trimmed = text.strip()

    # Count total characters
    total_chars = len(trimmed)

    # Count lines (sections)
    lines = [line for line in trimmed.split('\n') if line.strip()]
    line_count = len(lines)

    # Calculate remaining characters
    limit = 1000
    remaining = limit - total_chars

    return {
        'total_chars': total_chars,
        'line_count': line_count,
        'remaining': remaining,
        'is_valid': total_chars <= limit,
        'percentage': (total_chars / limit) * 100
    }


def format_output(stats):
    """Format and print the statistics."""
    status_emoji = '✓' if stats['is_valid'] else '✗'
    status = 'VALID' if stats['is_valid'] else 'EXCEEDS LIMIT'

    print('\n' + '=' * 50)
    print('  Suno Prompt Character Count')
    print('=' * 50)
    print(f"\nTotal Characters: {stats['total_chars']}/1000")
    print(f"Sections (lines): {stats['line_count']}")
    print(f"Remaining: {stats['remaining']} characters")
    print(f"Usage: {stats['percentage']:.1f}%")
    print(f"\nStatus: {status_emoji} {status}")

    if not stats['is_valid']:
        print(f"\n⚠️  OVER LIMIT BY {abs(stats['remaining'])} CHARACTERS")
        print('\nSuggestions:')
        print('  • Remove redundant adjectives')
        print('  • Use commas instead of "and"')
        print('  • Condense longer phrases')
        print('  • Remove least essential mood descriptors')

    print('\n' + '=' * 50 + '\n')

    return 0 if stats['is_valid'] else 1


def main():
    """Main execution."""
    if len(sys.argv) > 1:
        # Command line argument
        input_text = sys.argv[1]
    elif not sys.stdin.isatty():
        # Piped input
        input_text = sys.stdin.read()
    else:
        print('Usage: python count-prompt.py "your prompt text"', file=sys.stderr)
        print('   Or: echo "your prompt text" | python count-prompt.py', file=sys.stderr)
        sys.exit(1)

    stats = count_prompt(input_text)
    sys.exit(format_output(stats))


if __name__ == '__main__':
    main()
